sdk/python/kfp/dsl/_container_op.py

```diff
@@ -1010,6 +1013,19 @@ def __init__(
         """
 
         super().__init__(name=name, init_containers=init_containers, sidecars=sidecars, is_exit_handler=is_exit_handler)
+
+        if not DISABLE_REUSABLE_COMPONENT_WARNING and '--component_launcher_class_path' not in (arguments or []):
+            # The warning is suppressed for pipelines created using the TFX SDK.
+            warnings.warn(
+                "Please create reusable components instead of constructing ContainerOp instances directly."
+                " Reusable components are shareable, portable and have compatibility and support guarantees."
```

Comment:
Does deprecation warning mean container op will be removed in the future?

> have compatibility and support guarantees

Can you explain this part?

I think people typically don't want a layer of abstraction before thinking about reusability. Is the best option a component with no arguments in that case?

Reply:
>Does deprecation warning mean container op will be removed in the future?
Not necessary. But its role would be "task-like object that you get when creating an instance of component when compiling pipeline using KFP's compiler". So the users will still be interacting with `ContainerOp` objects, but they won't be creating those objects manually.

>>have compatibility and support guarantees
>Can you explain this part?

Components have better support throughout the system compared with raw ContainerOp.
* ContainerOps do not have inputs
* ContainerOps do not have types (so the artifacts in MLMD are untyped)
* ContainerOps are not supported in graph components
* There are no plans to support ContainerOps in TFX, but there are already multiple experimenta implementations of component support

Components have a more supported way forward compared to ContainerOp. There are already multiple experimental implementations of KFP component support for TFX, but it's not really possible to support ContainerOps there.

>I think people typically don't want a layer of abstraction before thinking about reusability.

I understand. But the equivalent components are the same number of lines/characters as ContainerOp code while providing more benefits. Also, I've answered dozens of issues from people using ContainerOps and many of those would not happen is they just used components. For example, with ContainerOp people are not thinking how they consume outputs, so they sometimes consume big outputs as values and then get runtime errors due to the data size.

>Is the best option a component with no arguments in that case?

I think so.

```python
hello_op = kfp.components.load_component_from_text('''
implementation:
  container:
    image: alpine
    command: ["echo", "Hello world"]
''')
```
instead of
```python
hello_op = kfp.dsl.ContainerOp(
    name="Hello",
    image="alpine",
    command=["echo", "Hello world"],
)
```




Is the best option a component with no arguments in that case?

Reply:
Thanks! I'm convinced. It would be great adding these to documentation.

So the plan is deprecating ContainerOp right now and removing it in the future?
SGTM

/lgtm

Reply:
>So the plan is deprecating ContainerOp right now and removing it in the future?

For now there are no plans to actually deprecate the class or remove it. Here is how it currently works:

KFP: component.yaml --load--> Callable --instantiate--> ContainerOp object
TFX: component.yaml --load--> Callable --instantiate--> BaseComponent object

So ContainerOp will still exist. You'll be able to access `ContainerOp.outputs`, `ContainerOp.container.*`, , `ContainerOp.set_display_name()`, `ContainerOp.after()`, `ContainerOp.apply()`, `ContainerOp.add_volume()` etc. It's still `ContainerOp`. Basically, we're only deprecating calling `ContainerOp.__init__`, but everything else remains.

Reply:
>It would be great adding these to documentation.

Good idea. I'm working on updating the documentation.

Reply:
Got it, sounds good

Reply:
> KFP: component.yaml --load--> Callable --instantiate--> ContainerOp object
Will this continue to be the case? We have code depending on manipulating the returned ContainerOp in `kfp.dsl.pipeline` annotated functions.